torchaudio/prototype/tacotron2.py

```diff
@@ -0,0 +1,847 @@
+            [prenet_dim, prenet_dim])
+
+        self.attention_rnn = nn.LSTMCell(
+            prenet_dim + encoder_embedding_dim,
+            attention_rnn_dim)
+
+        self.attention_layer = _Attention(
+            attention_rnn_dim, encoder_embedding_dim,
+            attention_hidden_dim, attention_location_n_filter,
+            attention_location_kernel_size)
+
+        self.decoder_rnn = nn.LSTMCell(
+            attention_rnn_dim + encoder_embedding_dim,
+            decoder_rnn_dim, True)
+
+        self.linear_projection = _LinearNorm(
+            decoder_rnn_dim + encoder_embedding_dim,
+            n_mel * n_frames_per_step)
+
+        self.gate_layer = _LinearNorm(
+            decoder_rnn_dim + encoder_embedding_dim, 1,
+            bias=True, w_init_gain='sigmoid')
+
+    def _get_initial_frame(self, memory: Tensor) -> Tensor:
+        r"""Gets all zeros frames to use as the first decoder input.
+
+        Args:
+            memory (Tensor): Encoder outputs (n_batch, text_lengths.max(), encoder_embedding_dim)
+
+        Returns:
+            decoder_input (Tensor): all zeros frames (n_batch, text_lengths.max(), n_mel * n_frames_per_step)
+        """
+
+        B = memory.size(0)
+        dtype = memory.dtype
+        device = memory.device
+        decoder_input = torch.zeros(
+            B, self.n_mel * self.n_frames_per_step, dtype=dtype, device=device)
+        return decoder_input
+
+    def _initialize_decoder_states(self,
+                                   memory: Tensor
+                                   ) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor]:
+        r"""Initializes attention rnn states, decoder rnn states, attention
+        weights, attention cumulative weights, attention context, stores memory
+        and stores processed memory.
+
+        Args:
+            memory (Tensor): Encoder outputs with shape (n_batch, text_lengths.max(), encoder_embedding_dim).
+
+        Returns:
+            attention_hidden (Tensor): Hidden state of the attention LSTM with shape (n_batch, attention_rnn_dim).
+            attention_cell (Tensor): Hidden state of the attention LSTM with shape (n_batch, attention_rnn_dim).
+            decoder_hidden (Tensor): Hidden state of the decoder LSTM with shape (n_batch, decoder_rnn_dim).
+            decoder_cell (Tensor): Hidden state of the decoder LSTM with shape (n_batch, decoder_rnn_dim).
+            attention_weights (Tensor): Attention weights with shape (n_batch, text_lengths.max()).
+            attention_weights_cum (Tensor): Cumulated attention weights with shape (n_batch, text_lengths.max()).
+            attention_context (Tensor): Context vector with shape (n_batch, encoder_embedding_dim).
+            processed_memory (Tensor): Processed encoder outputs
+                with shape (n_batch, text_lengths.max(), attention_hidden_dim).
+        """
+        B = memory.size(0)
+        MAX_TIME = memory.size(1)
```

Comment:
Can we use lower case `max_time`?
All capitalized looks like this is global variable.

Reply:
Thanks for pointing it out. It's addressed [here](https://github.com/pytorch/audio/pull/1621/commits/68b127377f8f8f8b6ae64c06ff4f6de885335d57#diff-8b8ec57192b9aaeb99a9ec5e709c977771b31b49fe7ddc6704c2128cf161a00eR502).